reasoner_pydantic/shared.py

```diff
@@ -10,15 +13,44 @@
 from .base_model import BaseModel
 from .utils import HashableSequence
 
-
+# TODO: potential add validation for structure of CURIE
 class CURIE(str):
     """Compact URI."""
 
 
+class KnowledgeType(str, Enum):
+    "Knowledge Type."
+
+    lookup = "lookup"
+    inferred = "inferred"
+
+
 class EdgeIdentifier(str):
     """Identifier for an edge in a knowledge graph"""
 
 
+class RecurrsiveAttribute(BaseModel):
+    """Attribute subattribute."""
+
+    attribute_type_id: CURIE = Field(..., title="type")
+    value: Any = Field(..., title="value")
+    value_type_id: Optional[CURIE] = Field(
+        None,
+        title="value_type_id",
+        nullable=True,
+    )
+    original_attribute_name: Optional[str] = Field(None, nullable=True)
+    value_url: Optional[str] = Field(None, nullable=True)
+    attribute_source: Optional[str] = Field(None, nullable=True)
+    description: Optional[str] = Field(None, nullable=True)
+    attributes: Optional[HashableSequence[RecurrsiveAttribute]] = Field(
+        None, nullable=True
+    )
+
+    class Config:
+        extra = "forbid"
+
+
 class SubAttribute(BaseModel):
```

Comment:
Do we need Attribute and SubAttribute anymore? Can we just use RecursiveAttribute in kgraph?

Reply:
I kept it mainly in case there was any instance in which the Attribute or SubAttribute object was handled directly. It just preserves functionality to utilize the RecurrsiveAttribute object to handle deeper levels.